File: source/RolXServer/RolXServer/Reports/WebApi/Resource/UserMonthReport.cs
// -----------------------------------------------------------------------
// <copyright file="UserMonthReport.cs" company="Christian Ewald">
// Copyright (c) Christian Ewald. All rights reserved.
// Licensed under the MIT license.
// See LICENSE.md in the project root for full license information.
// </copyright>
// -----------------------------------------------------------------------

using RolXServer.Users.WebApi.Resource;

namespace RolXServer.Reports.WebApi.Resource;

/// <summary>
/// The monthly report of a single user.
/// </summary>
public sealed record UserMonthReport(
    string Month,
    User User,
    IImmutableList<PartTimeSetting> PartTimeSettings,
    IImmutableList<BalanceCorrection> BalanceCorrections,
    TimeSpan Overtime,
    TimeSpan OvertimeDelta,
    TimeSpan Vacation,
    double? VacationDays,
    TimeSpan VacationDelta,
    double? VacationDeltaDays,
    IImmutableList<Domain.Model.WorkItemGroup> WorkItemGroups);

File: source/rolx-client/src/app/reports/core/user-month-report.ts
import { Duration, TransformAsDuration } from '@app/core/util/duration';
import { TransformAsIsoDate } from '@app/core/util/iso-date';
import { assertDefined } from '@app/core/util/utils';
import { WorkItemGroup } from '@app/reports/core/work-item-group';
import { BalanceCorrection } from '@app/users/core/balance-correction';
import { PartTimeSetting } from '@app/users/core/part-time-setting';
import { User } from '@app/users/core/user';
import { Type } from 'class-transformer';
import * as moment from 'moment';

export class UserMonthReport {
  @Type(() => User)
  user!: User;

  @TransformAsIsoDate()
  month!: moment.Moment;

  @Type(() => PartTimeSetting)
  partTimeSettings!: PartTimeSetting[];

  @Type(() => BalanceCorrection)
  balanceCorrections!: BalanceCorrection[];

  @TransformAsDuration()
  overtime!: Duration;

  @TransformAsDuration()
  overtimeDelta!: Duration;

  @TransformAsDuration()
  vacation!: Duration;

  @TransformAsDuration()
  vacationDelta!: Duration;

  vacationDays!: number | null;
  vacationDeltaDays!: number | null;

  @Type(() => WorkItemGroup)
  workItemGroups!: WorkItemGroup[];

  validateModel(): void {
    assertDefined(this, 'user');
    assertDefined(this, 'month');
    assertDefined(this, 'partTimeSettings');
    assertDefined(this, 'balanceCorrections');
    assertDefined(this, 'overtime');
    assertDefined(this, 'overtimeDelta');
    assertDefined(this, 'vacation');
    assertDefined(this, 'vacationDays');
    assertDefined(this, 'vacationDelta');
    assertDefined(this, 'vacationDeltaDays');
    assertDefined(this, 'workItemGroups');

    this.partTimeSettings.forEach((o) => o.validateModel());
    this.balanceCorrections.forEach((o) => o.validateModel());
    this.workItemGroups.forEach((o) => o.validateModel());
  }
}
